activerecord/lib/active_record/associations/join_dependency/join_association.rb

```diff
@@ -56,7 +56,7 @@ def join_constraints(foreign_table, foreign_klass, node, join_type, tables, scop
 
             klass_scope =
               if klass.current_scope
-                klass.current_scope.clone
+                remove_circular_references(klass.current_scope.clone)
```

Comment:
Join associations are a part of a join root. This means that join parts should prevent to construct another join root.

```diff
--- a/activerecord/lib/active_record/associations/join_dependency/join_association.rb
+++ b/activerecord/lib/active_record/associations/join_dependency/join_association.rb
@@ -56,7 +56,9 @@ def join_constraints(foreign_table, foreign_klass, node, join_type, tables, scop
 
             klass_scope =
               if klass.current_scope
-                klass.current_scope.clone
+                klass.current_scope.clone.tap { |scope|
+                  scope.joins_values = []
+                }
               else
                 relation = ActiveRecord::Relation.create(
                   klass,
```